clowder-framework/clowder
# Support DefaultAWSCredentialsProviderChain for passing in AWS S3 credentials

## Description


Fixes #200

### Problem
Clowder supports using S3 as backing storage for files, but this requires baking AWS credentials into our config file.

### Approach
* Fixed a bug where failed uploads would incorrectly decrement file + byte counts 
* Better support for [`DefaultAWSCredentialsProviderChain`](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html), so that other methods can be used to provide valid AWS credentials to the SDK
* NOTE: Due to a limitation in Play framework, `application.conf` manually detects and populates the necessary configs using environment variables. DO NOT specify environment variable credentials or static credentials in the config for scenarios where they should be ignored, as these explicitly-provided values will take precedence over other credentials.

### How to Test
1. Configure Clowder to use the `S3ByteStorageService`
    * (Minio or AWS S3) Provide a `clowder.s3.bucketName` in the Clowder configuration
    * (AWS S3 only) Provide a `clowder.s3.region` in the Clowder configuration
    * (Minio only) Provide a `clowder.s3.serviceEndpoint` in the Clowder configuration
2. Pass in your AWS credentials using a variety of methods:
    * (Minio or AWS S3) Environment variables (e.g. `AWS_ACCESS_KEY_ID ` + `AWS_SECRET_ACCESS_KEY ` OR `AWS_ACCESS_KEY` + `AWS_SECRET_KEY`)
    * (Minio or AWS S3) Java system properties (e.g. `aws.accessKeyId` + `aws.secretKey`)
    * (AWS S3 only) Web Identity Token (?)
    * (AWS S3 only) Credentials file at the default location (e.g. `~/.aws/credentials`)
    * (AWS S3 only) EC2 container service credentials (e.g. `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`)
    * (AWS S3 only) Instance profile credentials provided by AWS EC2 metadata service
3. Verify that Clowder starts up successfully and view the bucket in the AWS S3 or Minio management UI
    * You should see that the configured bucket has been created in S3
4. Create a dataset (if you haven't already), and upload a file to the dataset
5. Check the AWS S3 or Minio management UI again
    * You should see that an object has been created in the S3 bucket at the configured `clowder.s3.depth` (default: 3)

## Review Time Estimate


## Types of changes


## Checklist:
